libsignal-service/src/groups_v2/model.rs

```diff
@@ -0,0 +1,246 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Member")
+            .field("uuid", &self.uuid)
+            .field("role", &self.role)
+            .field("joined_at_revision", &self.joined_at_revision)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct PendingMember {
+    pub uuid: Uuid,
+    pub role: Role,
+    pub added_by_uuid: Uuid,
+    pub timestamp: u64,
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+pub struct RequestingMember {
+    pub uuid: Uuid,
+    pub profile_key: ProfileKey,
+    pub timestamp: u64,
+}
+
+impl PartialEq for RequestingMember {
+    fn eq(&self, other: &Self) -> bool {
+        self.uuid == other.uuid
+    }
+}
+
+impl fmt::Debug for RequestingMember {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RequestingMember")
+            .field("uuid", &self.uuid)
+            .field("timestamp", &self.timestamp)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum AccessRequired {
+    Unknown,
+    Any,
+    Member,
+    Administrator,
+    Unsatisfiable,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct AccessControl {
+    pub attributes: AccessRequired,
+    pub members: AccessRequired,
+    pub add_from_invite_link: AccessRequired,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Group {
+    pub title: String,
+    pub avatar: String,
+    pub disappearing_messages_timer: Option<Timer>,
+    pub access_control: Option<AccessControl>,
+    pub revision: u32,
+    pub members: Vec<Member>,
+    pub pending_members: Vec<PendingMember>,
+    pub requesting_members: Vec<RequestingMember>,
+    pub invite_link_password: Vec<u8>,
+    pub description: Option<String>,
+}
+
+#[derive(Clone, Debug)]
+pub struct GroupChanges {
+    pub editor: Uuid,
+    pub revision: u32,
+    pub changes: Vec<GroupChange>,
+}
+
+#[derive(Clone)]
+pub enum GroupChange {
+    NewMember(Member),
+    DeleteMember(Uuid),
+    ModifyMemberRole { uuid: Uuid, role: Role },
+    ModifyMemberProfileKey { uuid: Uuid, profile_key: ProfileKey },
+    // for open groups
+    NewPendingMember(PendingMember),
+    DeletePendingMember(Uuid),
+    PromotePendingMember { uuid: Uuid, profile_key: ProfileKey },
+    // when admin control is enabled
+    NewRequestingMember(RequestingMember),
+    DeleteRequestingMember(Uuid),
+    PromoteRequestingMember { uuid: Uuid, role: Role },
+    // group metadata
+    Title(String),
+    Avatar(String),
+    Timer(Option<Timer>),
+    Description(Option<String>),
+    AttributeAccess(AccessRequired),
+    MemberAccess(AccessRequired),
+    InviteLinkAccess(AccessRequired),
+    InviteLinkPassword(String),
+    AnnouncementOnly(bool),
+}
+
+impl fmt::Debug for GroupChange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
```

Comment:
What's the reason for not deriving here? If there's a legitimate reason, maybe document it as a comment, because I don't see it at first sight.

Reply:
🍝